compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -0,0 +1,402 @@
+import org.jetbrains.kotlin.backend.common.ir.copyTo
+import org.jetbrains.kotlin.backend.common.ir.copyTypeParametersFrom
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.irBlockBody
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.backend.jvm.ir.ScopedValueMappingTransformer
+import org.jetbrains.kotlin.backend.jvm.lower.inlineclasses.*
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.builders.declarations.buildConstructor
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrFunctionReferenceImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.defaultType
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+val jvmInlineClassPhase = makeIrFilePhase(
+    ::JvmInlineClassLowering,
+    name = "Inline Classes",
+    description = "Lower inline classes"
+)
+
+/**
+ * Adds new constructors, box, and unbox functions to inline classes as well as replacement
+ * functions and bridges to avoid clashes between overloaded function. Changes calls with
+ * known types to call the replacement functions.
+ *
+ * We do not unfold inline class types here. Instead, the type mapper will lower inline class
+ * types to the types of their underlying field.
+ */
+private class JvmInlineClassLowering(private val context: JvmBackendContext) : FileLoweringPass, ScopedValueMappingTransformer() {
+    private val manager = InlineClassManager()
+
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid()
+    }
+
+    override fun visitClass(declaration: IrClass): IrStatement {
+        scoped {
+            // The arguments to the primary constructor are in scope in the initializers of IrFields.
+            declaration.constructors.singleOrNull { it.isPrimary }?.let { primaryConstructor ->
+                manager.getReplacementFunction(primaryConstructor)?.let { replacement ->
+                    addMappings(replacement.valueParameterMap)
+                }
+            }
+
+            declaration.transformDeclarationsFlat { memberDeclaration ->
+                when (memberDeclaration) {
+                    is IrFunction ->
+                        transformFunctionFlat(memberDeclaration)
+                    else -> {
+                        memberDeclaration.accept(this, null)
+                        listOf(memberDeclaration)
```

Comment:
Please use `null` here (also consider converting `when` to `if`)

Reply:
Returning `null` from `transformDeclarationsFlat` will remove the declaration in question, which is not what we want.

I changed the `when` to an `if`, though. There used to be more cases. ;)

Reply:
Maybe I misunderstand something, but the kdoc says explicitly:
https://github.com/JetBrains/kotlin/blob/c67517cb2bc99844d5393a655ef264ce46d05b73/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/transform.kt#L32

Reply:
Huh, you're right. No idea why I thought otherwise. I'll change it momentarily.

Reply:
Done